### ISSUE TITLE
RR policy's connectivity status rules need improvement

### ISSUE THREAD

**Issue by AspirinSJL** [grpc/grpc#12969]
https://github.com/grpc/grpc/blob/master/src/core/ext/filters/client_channel/lb_policy/round_robin/round_robin.cc#L499

The current rules don't cover all the situations. As a result, the LB policy's status may not be updated when it should be. 

Since the change to the rules may affect the customers' view, how to improve the rules need discussion.

**Comment by AspirinSJL** [comment]
My idea is like below:

```
  /* In priority order. The first rule to match terminates the search (ie, if we
   * are on rule n, all previous rules were unfulfilled).
   *
   * 1) RULE: ANY subchannel is READY => policy is READY.
   *    CHECK: subchannel_list->num_ready > 0.
   *
   * 2) RULE: ANY subchannel is CONNECTING => policy is CONNECTING.
   *    CHECK: sd->curr_connectivity_state == CONNECTING.
   *
   * 3) RULE: ALL subchannels are IDLE => policy is IDLE.
   *    CHECK: subchannel_list->num_idle == subchannel_list->num_subchannels.
   *
   * 4) RULE: ALL subchannels are either IDLE, SHUTDOWN or TRANSIENT_FAILURE but
   *    at least one subchannel is not IDLE => policy is TRANSIENT_FAILURE.
   *    (This covers all the rest situations.)
   *    CHECK: (subchannel_list->num_shutdown +
   *            subchannel_list->num_transient_failures +
   *            subchannel_list->num_idle ==
   *            subchannel_list->num_subchannels) &&
   *           (subchannel_list->num_shutdown +
   *            subchannel_list->num_transient_failures > 0)
   */
```

**Comment by markdroth** [comment]
This is related to the issue of improving the connectivity state semantics spec:

https://github.com/grpc/grpc/blob/master/doc/connectivity-semantics-and-api.md

**Comment by AspirinSJL** [comment]
Resolved via #12829, #13654, and #13661.
